src/tests/end2end/CopyTests.cpp

```diff
@@ -0,0 +1,502 @@
+    BufferSpec pixelBuffer = MinimumBufferSpec(1, 1);
+    DoTest({ kWidth, kHeight, 0, 0, 1, 1, 0 },                    pixelBuffer);
+    DoTest({ kWidth, kHeight, kWidth - 1, 0, 1, 1, 0 },           pixelBuffer);
+    DoTest({ kWidth, kHeight, 0, kHeight - 1, 1, 1, 0 },          pixelBuffer);
+    DoTest({ kWidth, kHeight, kWidth - 1, kHeight - 1, 1, 1, 0 }, pixelBuffer);
+    DoTest({ kWidth, kHeight, kWidth / 3, kHeight / 7, 1, 1, 0 }, pixelBuffer);
+    DoTest({ kWidth, kHeight, kWidth / 7, kHeight / 3, 1, 1, 0 }, pixelBuffer);
+}
+
+// Test that copying regions with 256-byte aligned sizes works
+TEST_P(CopyTests_T2B, TextureRegionAligned) {
+    constexpr uint32_t kWidth = 256;
+    constexpr uint32_t kHeight = 128;
+    for (unsigned int w : {64, 128, 256}) {
+        for (unsigned int h : { 16, 32, 48 }) {
+            DoTest({ kWidth, kHeight, 0, 0, w, h, 0 }, MinimumBufferSpec(w, h));
+        }
+    }
+}
+
+// Test that copying regions without 256-byte aligned sizes works
+TEST_P(CopyTests_T2B, TextureRegionUnaligned) {
+    constexpr uint32_t kWidth = 256;
+    constexpr uint32_t kHeight = 128;
+    for (unsigned int w : {13, 63, 65}) {
+        for (unsigned int h : { 17, 19, 63 }) {
+            DoTest({ kWidth, kHeight, 0, 0, w, h, 0 }, MinimumBufferSpec(w, h));
+        }
+    }
+}
+
+// Test that copying mips with 256-byte aligned sizes works
+TEST_P(CopyTests_T2B, TextureMipAligned) {
+    constexpr uint32_t kWidth = 256;
+    constexpr uint32_t kHeight = 128;
+    for (unsigned int i = 1; i < 4; ++i) {
+        DoTest({ kWidth, kHeight, 0, 0, kWidth >> i, kHeight >> i, i }, MinimumBufferSpec(kWidth >> i, kHeight >> i));
+    }
+}
+
+// Test that copying mips without 256-byte aligned sizes works
+TEST_P(CopyTests_T2B, TextureMipUnaligned) {
+    constexpr uint32_t kWidth = 259;
+    constexpr uint32_t kHeight = 127;
+    for (unsigned int i = 1; i < 4; ++i) {
+        DoTest({ kWidth, kHeight, 0, 0, kWidth >> i, kHeight >> i, i }, MinimumBufferSpec(kWidth >> i, kHeight >> i));
+    }
+}
+
+// Test that copying with a 512-byte aligned buffer offset works
+TEST_P(CopyTests_T2B, OffsetBufferAligned) {
+    constexpr uint32_t kWidth = 256;
+    constexpr uint32_t kHeight = 128;
+    for (unsigned int i = 0; i < 3; ++i) {
+        BufferSpec bufferSpec = MinimumBufferSpec(kWidth, kHeight);
+        uint32_t offset = 512 * i;
+        bufferSpec.size += offset;
+        bufferSpec.offset += offset;
+        DoTest({ kWidth, kHeight, 0, 0, kWidth, kHeight, 0 }, bufferSpec);
+    }
+}
+
+// Test that copying without a 512-byte aligned buffer offset works
+TEST_P(CopyTests_T2B, OffsetBufferUnaligned) {
+    if (IsD3D12()) {
+        printf("TODO(enga@google.com): Test skipped because unaligned buffer offsets not supported in D3D12\n");
+        return;
+    }
+    constexpr uint32_t kWidth = 128;
+    constexpr uint32_t kHeight = 128;
+    for (uint32_t i = kBytesPerTexel; i < 512; i += kBytesPerTexel * 3) {
```

Comment:
That looks like a lot of iterations if it takes much longer than the other tests, can we have the increment be bigger?